CAIPs/caip-217.md

```diff
@@ -0,0 +1,99 @@
+mechanisms, for example:
+- A JSON-RPC protocol for persisting and synchronizing authorized sessions
+  ([CAIP-25][])
+- Routing individual RPC commands to an authorized network ([CAIP-27][])
+
+## Motivation
+
+The layering of today's cryptographic and decentralized systems favors
+loosely-coupled combinations of protocols (representated in the CAIPs model as
+[namespaces][]), instances or consensus-communities within those protocols
+(addressed in the CAIPs model as [CAIP-2][] URNs), and sets of supported RPC
+methods and notifications used in those namespaces. Bundling all of these into
+an object facilitates unambiguous authorization schemes, including progressive
+authorization patterns, feature discovery, authority negotiation (See
+[CAIP-211][]) and delegations.
+
+## Specification
+
+An authorization scope is represented in JSON as a string that expresses its
+scope and a JSON object defining the bundle of properties authorized there. When
+embedded in any other JSON context (including the `params` of a JSON-RPC
+message), the object MUST be expressed as the value of a property named by the
+scope string. 
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in
+uppercase in this document are to be interpreted as described in [RFC
+2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Definition
+
+The syntax is as follows:
+
+```jsonc
+
+scope: string 
+
+scopeObject: {
+   *scopes: [(chainId)+],
+   methods: [(method_name)+],
+   notifications: [(notification_name)+],
+   *accounts: [(account_id)+]
+   *rpcDocuments: [(rpcDocument)+],
+   *rpcEndpoints: [(rpcEndpoint)+]
+}
+```
+
+Where:
+
+- `string` (conditional) = A namespace identifier string registered in the CASA [namespaces][] registry to authorize multiple chains with identical properties OR a single, valid [CAIP-2][] identifier, i.e., a specific `chain_id` within a namespace.
+- `scopes` (conditional) = An array of 0 or more [CAIP-2][] `chainId`s. Setting multiple `scopes` is equivalent to making multiple copies of the `scopeObject` for each member of `scopes`, simply in compact form.
+  - This property MUST NOT be present if the object is already scoped to a single `chainId` in the string value above.
+  - This property MUST NOT be present if the scope is an entire [namespace][namespaces] in which `chainId`s are not defined.
+  - This property MAY be present if the scope is an entire [namespace][namespaces] in which `chainId`s are defined.
+- `methods` = An array of 0 or more JSON-RPC methods that an application can call on the agent and/or an agent can call on an application.
+- `notifications` = An array of 0 or more JSON-RPC notifications that an application send to or expect from the agent.
+- `accounts` (optional) = An array of 0 or more [CAIP-10][] identifiers, each valid within the scope of authorization.
```

Comment:
How are we addressing account matching for namespaced scopes? Also shouldn't this be 1 or more CAIP-10 identifiers (in the response)?

Reply:
1 or more in eip155, but not necessarily in all namespaces (for example, the hypothetical `wallet` namespace!). I think we'd need to merge this and then open a PR to discuss in two weeks adding `eip155/caip25.md` to namespaces which adds the "always 1 account in every `sessionScopes` member keyed to eip155 or to an eip155 CAIP-2..." rule :D

Reply:
Ok, I'm fine with keeping it at 0 or more, but we do need to include some sort of language and logic around account matching, we talked about how the namespace scope was a shorthand for chain specific scopes and how the `accounts` logic doesn't match up with that for that exact reason. What if a chain returns more than one account, what if a chain doesn't return any accounts? We have no idea how to match based on an array. I think omitting this leaves this CAIP incomplete especially when it's introducing this concept of `accounts`. We're making the incorrect assumption that namespaces share accounts with the current spec.

Reply:
right-- I think maybe the only implicit "matching" logic was the requirement that `accounts` are "valid" in the given scope of a `scopeObject`, but you're describing a corner-case of that squishy "validity". For example:

```
"eip155": {
...
   "scopes": ["chainid1", "chainid2", "chainid3"],
   "accounts"["eip155:chainid1:acct1", "eip155:chainid2:acct2", "eip155:chainid3:acct3"]
...
}
```
all three accounts are valid in `eip155` but `acct3` is valid only in `chainid3`, so if the `scopes` array is actually a shorthand for condensing 3 distinct caip2-specific scopes, all 3 of them have 2 invalid accounts in their `accounts` array, perhaps?

I think maybe we're trying to solve an eip155 problem at the CAIP layer, and the desired matching behavior for this corner-case might best be specified by the `eip155/caip25.md` profile? Or do we need to expand the concept (or the syntax) of `accounts` for all namespaces?  Even if namespaces like polkadot can "convert" a universal address to a CAIP2-specific address and thus sidestep the problem of matching `accounts` to `scopes`, Solana and some other namespaces have an additional matching problem, where each address exposed needs to be matched to a public key...

Reply:
So can we merge this for the sake of updating `main` (and the .org website) and return to this issue later, after we've tried writing the `eip155/caip25.md` profile? It might be worth detailing how to handle "invalid" `accounts` (drop? ignore? drop-or-ignore?) but it's hard to specify until we have an "end to end example" , i.e. a stable CAIP-25 and a profile defining validity in namespace-specific ways :D

Reply:
yes, i'm fine with leaving this part as is.

Reply:
Closing in the hopes that [this commit](https://github.com/ChainAgnostic/CAIPs/pull/217/commits/8a92f2b79f34ac64aaeb0eab6b29b3ce0ca55e80) leaves validity to be defined in profiles of the spec